src/Native/System.Native/pal_networkchange.cpp

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_config.h"
+
+#if HAVE_LINUX_RTNETLINK_H
+#include "pal_networkchange.h"
+#include "pal_types.h"
+#include "pal_utilities.h"
+
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <net/if.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#pragma clang diagnostic ignored "-Wcast-align" // NLMSG_* macros trigger this
+
+extern "C" int32_t CreateNetworkChangeListenerSocket()
+{
+    sockaddr_nl sa = { };
+    sa.nl_family = AF_NETLINK;
+    sa.nl_groups = RTMGRP_LINK | RTMGRP_IPV4_IFADDR;
+    int32_t sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (bind(sock, reinterpret_cast<sockaddr*>(&sa), sizeof(sa)) != 0)
+    {
+        return -1;
+    }
+    return sock;
+}
+
+extern "C" int32_t CloseNetworkChangeListenerSocket(int32_t socket)
```

Comment:
In managed code, this is typed as returning an `Error`, but it doesn't: it just returns an `int`. This should look something like:

```
extern "C" Error CloseNetworkChangeListenerSocket(int32_t socket)
{
    int err = close(socket);
    return err == 0 ? PAL_SUCCESS : ConvertErrorPlatformToPal(errno);
}
```

Other functions that intend to return `Error` values should be similar.